ESQL: add Arrow dataframes output format

swallez opened the pull request:
Adds support for [Apache Arrow](https://arrow.apache.org/docs/index.html)'s [streaming format](https://arrow.apache.org/docs/format/Columnar.html#ipc-streaming-format) as a response for ES|QL. It triggers based on the `Accept` header or the `format` request parameter.

Arrow has implementations in every mainstream language and is a backend of the Python [Pandas](https://pandas.pydata.org/) library, which is extremely popular among data scientists and data analysts. Arrow's streaming format has also become the de facto standard for dataframe interchange. It is an efficient binary format that allows zero-cost deserialization by adding data access wrappers on top of memory buffers received from the network.

This PR builds on the experiment made by @nik9000 in PR #104877

**Features/limitations**

- all ES|QL data types are supported
- multi-valued fields are not supported
- fields of type `_source` are output as JSON text in a `varchar` array. In a future iteration we may want to offer the choice of the more efficient CBOR and SMILE formats.

**Technical details**

Arrow comes with its own memory management to handle vectors with direct memory, reference counting, etc. We don't want to use this as it conflicts with Elasticsearch's own memory management.

We therefore use the Arrow library only for the metadata objects describing the dataframe schema and the structure of the streaming format. The Arrow vector data is produced directly from ES|QL blocks.

**Future work**

Although it's already quite featured, there are some improvements to be done, beyond the limitations outlined above:
- we should be able to significantly speed things up by copying directly the ES|QL vector's memory buffer instead of reading each position in a loop
- ES|QL creates small pages when there are some `null` values. We should aggregate small pages into larger Arrow record batches to remove the overhead of batch headers.

**Tests**

The unit tests cover all data types and random combinations of fields and sparse/dense/empty vectors. In addition to that, some manual tests have been done in Python, Rust and Java using the standard Arrow libraries.

Below is a Python example taken from PR #104877. The last 3 lines are where the Arrow format is used: read the Arrow stream, wrap it in a Pandas dataframe, and plot the data.

```python
import base64
import urllib3
import pyarrow as pa
import matplotlib.pyplot as plt
import pandas as pd
plt.close("all")
timeout = urllib3.Timeout(connect=2.0, read=300.0)
http = urllib3.PoolManager(timeout=timeout)

def basic(login, password):
  encoded = base64.b64encode((login + ':' + password).encode()).decode()
  return "Basic %s" % encoded

resp = http.request(
  "POST",
  "http://elastic:password@localhost:9200/_query?format=arrow",
  headers = {
    "authorization": basic("elastic", "password")
  },
  json = {
    "query": """
      FROM nyc_taxis 
    | WHERE trip_distance > 0 AND trip_distance < 100 AND fare_amount > 1
    | STATS min=MIN(fare_amount), avg=AVG(fare_amount), max=MAX(fare_amount) BY pickup_datetime=DATE_TRUNC(1 DAY, pickup_datetime)
    | EVAL LOG10(min), LOG10(max), LOG10(avg)
    | DROP min, max, avg
    | SORT pickup_datetime
    | LIMIT 100000"""
  })

if resp.status != 200:
  print("failed: %s" % resp.data)
else:
    with pa.ipc.open_stream(resp.data) as reader:
        df = reader.read_pandas()
        df.plot.line(x="pickup_datetime")
```

swallez committed:
ESQL: add Arrow dataframes output format

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

nik9000 commented:
This has been on my review list for a few days. I'm just on vacation a bunch of days. It's still on the list. sorry for the delay!

nik9000 commented on x-pack/plugin/esql/arrow/src/main/java/org/elasticsearch/xpack/esql/arrow/ArrowResponse.java:
```diff
@@ -0,0 +1,376 @@
+    @Override
+    public boolean isLastPart() {
+        // Even if sent in chunks, the entirety of ESQL data is available, so it's single (chunked) part
+        return true;
```
In a follow-up it'd be cool to change this. Arrow's IPC seems designed to work properly with streaming these results. We haven't plumbed that out of Elasticsearch though. And it's hard! Certainly not for this PR.

nik9000 replied:
Sorry, I meant change it so we don't have to wait for all of the results to arrive. Nothing to do for now, there's lots of plumbing to make it work.

nik9000 commented on x-pack/plugin/esql/arrow/src/main/java/org/elasticsearch/xpack/esql/arrow/BlockConverter.java:
```diff
@@ -0,0 +1,477 @@
+    protected BlockConverter(String esqlType, Types.MinorType minorType) {
+        // Add the exact ESQL type as field metadata
+        var meta = Map.of("elastic:type", esqlType);
+        this.fieldType = new FieldType(true, minorType.getType(), null, meta);
```
:+1:

nik9000 commented on x-pack/plugin/esql/arrow/src/main/java/org/elasticsearch/xpack/esql/arrow/BlockConverter.java:
```diff
@@ -0,0 +1,477 @@
+    }
+
+    protected final <T extends Block> T castBlock(Block block, Class<T> clazz) {
+        if (clazz.isAssignableFrom(block.getClass())) {
```
This is pretty much what doing `(DoubleBlock) block` does but it works with the runtime `Class` object. I think it's probably ok to just do the cast without the method.

swallez replied:
Removed in c77c2459e2b7cec24793b5a664f42615f281d1dd to use a regular cast. This was useful in development, but this should never fail as unit tests test all field types.

nik9000 commented on x-pack/plugin/esql/arrow/src/main/java/org/elasticsearch/xpack/esql/arrow/BlockConverter.java:
```diff
@@ -0,0 +1,477 @@
+                // FIXME: are these parameters ok here?
+                // FIXME: should the resulting block be closed after use?
+                new NoopCircuitBreaker("esql-arrow"),
+                BigArrays.NON_RECYCLING_INSTANCE
```
This is fine for a prototype but it's a blocker to stop switch. There are three choices I think that are ok:
1. Push the converter into the superclass and do it optionally, right as you dump into the wire. This is probably the best way because it involves the least copying and memory management. It might be the slowest because the converter will go megamorphic pretty quick. But it's shoveling a fair number of bits around so it's probably fine. Probably.
2. Convert the whole block up front in a tight loop with memory management.
3. Option 1 but with the generics reified out by hand - java doesn't do it for you so you need to do it with your own fingers. Or, in our case, we'd do it with stringtemplate. That's almost certainly the choice that's fastest at runtime. But I don't know that it matters for this.

Either way, this is fine for now.

swallez replied:
I don't know much about memory management in ES. Is there some explanation somewhere or some reference code I could get inspiration from?

nik9000 replied:
It's more memory management int he compute engine than the rest of ES. The short version is that you can use `BlockFactory` to make a builder like:
```
try (IntBlockBuilder builder = block.blockFactory().newIntBlockBuilder(block.getPositionCount())) {
  for... { builder.appendInt(i); }
  try (IntBlock block = builder.build()) {
    super.render(block);
  }
}
```
That'll account for the new block using the same circuit breakers as the original block. It *could* be fairly large.

Something like `RoundIntEvaluator` does something similar.
}

swallez replied:
Added 54fa11f52dba3332abffd2cf06e5d95a47283af2 to reuse the source block's factory and inherits its circuit breaker.

nik9000 commented on x-pack/plugin/esql/arrow/src/main/java/org/elasticsearch/xpack/esql/arrow/BlockConverter.java:
```diff
@@ -0,0 +1,477 @@
+                        if (bytes.length != 0) {
+                            bytes = valueConverter.apply(bytes, scratch);
+                        }
+                        builder.appendBytesRef(bytes);
```
You are probably better off copying the bytes to a single `BytesRefBuilder` scratch and then passing *that* to the converter to modify in place. That way it doesn't have to allocate the result.

One interesting thing I've realized is that in most cases it's safe to just directly modify the returned `bytes` but not in all cases. Which is kind of lame.

nik9000 reviewed (commented):
I left a few comments, but I think the most important merge blockers are:

1.  Some kind of integration test. I guess it'd be something that uses the arrow ipc library directly in java to read and write some things.
2. Only allowing this in SNAPSHOT builders. It's super paranoid behavior, but I think it's warranted.

What should we do about multivalued results, do you think? By the time we go to serialize these we're past the point of adding warning headers. At least for now, we could silently drop them. Or maybe log a warning in the server log.

nik9000 commented:
> Only allowing this in SNAPSHOT builds. It's super paranoid behavior, but I think it's warranted.

@swallez and I talked about this some offline and I'm convinced it's fine to allow this in non-snapshot builds. It's an extra layer of paranoia that in this case isn't buying us anything. It has really useful side effects when we are making inter-node communication changes, but we aren't doing that here.

swallez committed:
Change status from tech preview to experimental

swallez committed:
Cleanup exception handling

swallez commented:
@nik9000 I addressed your comments. Regarding the main points:

> Some kind of integration test. I guess it'd be something that uses the arrow ipc library directly in java to read and write some things.

In `ArrowResponseTests.toArrowVectors()` we serialize the ESQL response to a byte array and read it with Arrow's `ArrowStreamReader`. There is no network involved, but it seemed fine to me as we want to test serialization and deserialization with standard Arrow libraries.

> Only allowing this in SNAPSHOT builders. It's super paranoid behavior, but I think it's warranted.

Since it's really peripheral (it's an output format users have to choose explicitly and involves no communication between cluster nodes), it seems overly paranoid to me 😉 

> What should we do about multivalued results, do you think? By the time we go to serialize these we're past the point of adding warning headers. At least for now, we could silently drop them. Or maybe log a warning in the server log.

I did not look at it currently and considered them as single-valued blocks. Looking at the implementation this is actually wrong, since we will flatten all values up to the size of the page. I'll see if I can add them quickly. Otherwise we can either:
1. drop multivalued fields from the output as you suggest,
2. fail early saying that they're not supported yet (it's experimental).

I'd rather go for 2/ which is less surprising for users (i.e. no "but where is my field?")

swallez committed:
Fail hard when there are multi-valued blocks

swallez commented:
In c7cd12d I added a check to fail hard with a meaningful message when we encounter a multivalued block (and associated test).

nik9000 commented:
> I'd rather go for 2/ which is less surprising for users (i.e. no "but where is my field?")

:+1:

The reason I'm keen on involving the network isn't really the network, it's to get a view of how this all looks from the user's perspective with netty involved. For example, throwing your exception there should make sure to come back as an error, but if we've already started the response it won't quite work right. Mostly I'm interested in the error cases and maybe like 1 success case.

As a follow-up I can try and adapt the csv-spec tests to use array which gives us great coverage without having to maintain much more stuff.

nik9000 commented on x-pack/plugin/esql/arrow/build.gradle:
```diff
@@ -0,0 +1,61 @@
+  implementation('org.apache.arrow:arrow-vector:16.1.0')
+  implementation('org.apache.arrow:arrow-format:16.1.0')
+  implementation('org.apache.arrow:arrow-memory-core:16.1.0')
+  implementation('org.checkerframework:checker-qual:3.42.0')
```
:+1:

Cool. Ignore it.

nik9000 commented on x-pack/plugin/esql/arrow/src/main/java/org/elasticsearch/xpack/esql/arrow/ValueConversions.java:
```diff
@@ -0,0 +1,80 @@
+                return scratch;
+            }
+        }
+        return value;
```
It's fine to keep it where it is.

swallez committed:
Remove fancy block downcast. Wrong cast will be checked by unit tests

swallez committed:
Add test to ensure a null uses one position in the block

swallez commented on x-pack/plugin/esql/arrow/src/main/java/org/elasticsearch/xpack/esql/arrow/BlockConverter.java:
```diff
@@ -0,0 +1,477 @@
+
+                // TODO could we "just" get the memory of the array and dump it?
+                int count = block.getPositionCount();
+                for (int i = 0; i < count; i++) {
```
I added explicit rejection of multi-valued arrays (for now) in c7cd12d026558c73b1c847856fcf82551bbc9a16 so this is ok as we only have single-valued arrays.

Null is handled, based on the observation that a null value uses a position and its value is zero in number blocks. I don't know if this part of the contract (I added a test for it), but it should, so that we can optimize processing of arrays where `block.mayHaveMultivaluedFields() == true`.

The validity bitmap is also output in `BlockConverter.accumulateValidity()`.

That will of course have to be revisited when handling mutl-values.

swallez committed:
Create transformed byteref blocks using the original block's factory

nik9000 commented on x-pack/plugin/esql/arrow/src/main/java/org/elasticsearch/xpack/esql/arrow/BlockConverter.java:
```diff
@@ -0,0 +1,454 @@
+        @Override
+        public void convert(Block b, List<ArrowBuf> bufs, List<BufWriter> bufWriters) {
+            BytesRefBlock block = (BytesRefBlock) b;
+            BytesRefBlock transformed = transformValues(block);
```
I'm going to to try and write a test that fails without this.

swallez committed:
Add integration test

swallez commented:
@nik9000 I added an integration test in 1b75af7df08ac336006ee011506058dcfe28a118 - is this the kind of integration test you're expecting? Are there some additional checks that need to be done regarding memory management?

swallez committed:
Fix imports

nik9000 commented:
> @nik9000 I added an integration test in [1b75af7](https://github.com/elastic/elasticsearch/commit/1b75af7df08ac336006ee011506058dcfe28a118) - is this the kind of integration test you're expecting? Are there some additional checks that need to be done regarding memory management?

Bah that's pretty much what I was going to add. I just didn't see it. sorry. But, yeah, I do think it needs one that hits the types that need the transform.

I think a neat followup is to try and write a test that'll run the CSV tests fetching over arrow. But that's kind of finicky to get in, especially without multivalued fields. That'll get all the types and everything. But for now I think doing a test with some of the transformed types is good.

swallez committed:
Add integration test for data transformed during output

swallez commented:
Added integration tests for values and ip addresses (both are transformed on serialization) in 00aa730

nik9000 commented on x-pack/plugin/esql/arrow/src/main/java/org/elasticsearch/xpack/esql/arrow/BlockConverter.java:
```diff
@@ -0,0 +1,451 @@
+                        builder.appendNull();
+                    } else {
+                        BytesRef bytes = block.getBytesRef(i, scratch);
+                        if (bytes.length != 0) {
```
Now that I think about it, I think this should apply to empty `BytesRef` values. Let the converter decide what to do.

nik9000 reviewed (commented):
I have discovered a small :bug:. I'll push some changes to your test in a bit and maybe the bug fix. Checking.

nik9000 committed:
Track memory on conversion

nik9000 commented:
@swallez I figured out my issue. I was worried about that conversion code not cleaning up the memory tracking. And it wasn't. But your integration test didn't catch it because it was missing an explicit check for memory tracking. Now, what that isn't part of every test, I'm not sure. But it isn't. I had to opt it in. And when I did it started failing. So I added the tracking, just a `try` block, and broke the test up a bit so I could make sure it was just the transform.

Now I feel much better.

I'm going to give this one more read and then approve it.

elasticsearchmachine commented:
Hi @swallez, I've created a changelog YAML for you.

swallez committed:
Update docs/changelog/109873.yaml

swallez committed:
Merge remote-tracking branch 'upstream/main' into esql-arrow-output

nik9000 committed:
Fix merge

nik9000 commented:
@swallez I knew the change that broke the build for you so I pushed a fix.

nik9000 commented:
Pretty pictures you can make with pandas
![image](https://github.com/elastic/elasticsearch/assets/215970/f4b76cf6-abf9-4a71-aab7-dabbaebb7fc7)


![image](https://github.com/elastic/elasticsearch/assets/215970/db5f384b-51e0-4f35-9161-105f4f7fb7a3)
